stdlib/public/Concurrency/TaskGroup.swift

```diff
@@ -588,35 +572,70 @@ public struct ThrowingTaskGroup<ChildTaskResult: Sendable, Failure: Error> {
   ///         collected += value
   ///     }
   ///
-  /// ### Thread-safety
-  /// Please note that the `group` object MUST NOT escape into another task.
-  /// The `group.next()` MUST be awaited from the task that had originally
-  /// created the group. It is not allowed to escape the group reference.
+  /// If the next child task throws an error
+  /// and you propagate that error from this method
+  /// out of the body of a `TaskGroup.withThrowingTaskGroup(of:returning:body:)` call,
+  /// then all remaining child tasks in that group are implicitly canceled.
   ///
-  /// Note also that this is generally prevented by Swift's type-system,
-  /// as the `add` operation is `mutating`, and those may not be performed
-  /// from concurrent execution contexts, such as child tasks.
+  /// - Returns: The value returned by the next child task that completes.
   ///
-  /// ### Ordering
-  /// Order of values returned by next() is *completion order*, and not
-  /// submission order. I.e. if tasks are added to the group one after another:
+  /// - Throws: The error thrown by the next child task that completes.
+  ///
+  /// - SeeAlso: `nextResult()`
+  public mutating func next() async throws -> ChildTaskResult? {
+    return try await _taskGroupWaitNext(group: _group)
+  }
+
+  /// Wait for the next child task to complete,
+  /// and return a result containing either
+  /// the value that the child task returned or the error that it threw.
+  ///
+  /// The values returned by successive calls to this method
+  /// appear in the order that the tasks *completed*,
+  /// not in the order that those tasks were added to the task group.
+  /// For example:
   ///
   ///     group.spawn { 1 }
   ///     group.spawn { 2 }
   ///
-  ///     print(await group.next())
-  ///     /// Prints "1" OR "2"
+  ///     print(await group.nextResult())
+  ///     // Prints either "2" or "1"
   ///
-  /// ### Errors
-  /// If an operation added to the group throws, that error will be rethrown
-  /// by the next() call corresponding to that operation's completion.
+  /// If there aren't any pending tasks in the task group,
+  /// this method returns `nil`,
+  /// which lets you write like the following
+  /// to wait for a single task to complete:
+  ///
+  ///     if let first = try await group.next() {
+  ///        return first
+  ///     }
+  ///
+  /// Wait and collect all group child task completions:
+  ///
+  ///     while let first = try await group.next() {
+  ///        collected += value
+  ///     }
+  ///     return collected
+  ///
+  /// Awaiting on an empty group results in the immediate return of a `nil`
+  /// value, without the group task having to suspend.
+  ///
+  /// It is also possible to use `for await` to collect results of a task groups:
+  ///
+  ///     for await try value in group {
+  ///         collected += value
+  ///     }
```

Comment:
most of this comment is about `next()` now about `nextResult()`; let's only do a small section on:

```
guard let result = await group.nextResult() else {
  return // no task to wait on
}

switch result { 
case .success(let value): print(value)
case .failure(let error): print("oh no: \(error)")
}
```

maybe?

People will be using the `for await` and `next()` mostly